# Result Analysis

Setting up useful variables and external libraries

In [1]:
#%pdb on

import os
import pandas as pd

pth = '<score-charts-pth>'

## Data Understanding 

This table shows an example of game result.  

In [2]:
score_file = '<score-xls-file>' # Ex: 'score-nonstat_ne0.2_ge0.2_pg0.1_pa1.0.xls'
df = pd.read_excel(os.path.join(pth, ), header=0, index_col=0)
df

Exp3   FreqHist  Frequentist       Nash   NoRegret  \
Exp3               NaN  43.460000    85.726667  73.430000  73.623333   
FreqHist     56.540000        NaN    91.090000  73.443333  71.020000   
Frequentist  14.273333   8.910000          NaN  18.913333  35.160000   
Nash         26.570000  26.556667    81.086667        NaN  50.543333   
NoRegret     26.376667  28.980000    64.840000  49.456667        NaN   
Reply-Last   18.210000   8.460000    63.406667  35.716667  41.686667   
ReplyHist    55.333333  50.510000    90.453333  71.486667  71.280000   
Skynet       33.586667  50.086667    68.200000  74.290000  73.836667   
UCB1-Tuned   53.783333  48.596667    70.893333  74.746667  74.406667   
Xelnaga      28.286667  50.173333    91.700000  54.280000  54.980000   
e-Greedy     51.540000  45.493333    84.776667  64.230000  64.653333   
e-Nash       25.793333  32.260000    81.586667  43.633333  47.906667   

             Reply-Last  ReplyHist     Skynet  UCB1-Tuned    Xelnaga  \
Exp3          81.790000  44.666667  66.413333   46.216667  71.713333   
FreqHist      91.540000  49.490000  49.913333   51.403333  49.826667   
Frequentist   36.593333   9.546667  31.800000   29.106667   8.300000   
Nash          64.283333  28.513333  25.710000   25.253333  45.720000   
NoRegret      58.313333  28.720000  26.163333   25.593333  45.020000   
Reply-Last          NaN   8.726667  31.886667   25.146667   8.260000   
ReplyHist     91.273333        NaN  52.720000   50.910000  50.383333   
Skynet        68.113333  47.280000        NaN   29.393333  76.100000   
UCB1-Tuned    74.853333  49.090000  70.606667         NaN  75.233333   
Xelnaga       91.740000  49.616667  23.900000   24.766667        NaN   
e-Greedy      82.376667  44.476667  66.450000   45.520000  68.980000   
e-Nash        69.110000  29.003333  27.406667   26.483333  38.770000   

              e-Greedy     e-Nash  
Exp3         48.460000  74.206667  
FreqHist     54.506667  67.740000  
Frequentist  15.223333  18.413333  
Nash         35.770000  56.366667  
NoRegret     35.346667  52.093333  
Reply-Last   17.623333  30.890000  
ReplyHist    55.523333  70.996667  
Skynet       33.550000  72.593333  
UCB1-Tuned   54.480000  73.516667  
Xelnaga      31.020000  61.230000  
e-Greedy           NaN  64.823333  
e-Nash       35.176667        NaN

The line/column Exp3/FreqHist has the value 43.46. This represents the percentage of winning Exp3 has over FreqHist on 1000 rounds of a particular simulation.

The result file has the defined configuration on its name. The file is intepreted as follows:

- `nonstat`: represents if the game is stationary or non-stationary (possible values are `nonstat` and `stat`);
- `ne#`: $\varepsilon$ value of $\varepsilon$-Nash (`ne0.2` means $\varepsilon = 0.2$);
- `ge#`: $\epsilon$ value of $\epsilon$-Greedy;
- `pg#` and `pa#`: $\gamma$ and $\alpha$ of Exp3, respectively.

## Analysing the Data

Defining helper functions

In [3]:
def mean_players(xls_file, players):
    df = pd.read_excel(xls_file, header=0, index_col=0)
    return {(pname + '_mean'): df.loc[pname].mean() for pname in players}

In [4]:
def sim_conf(xls_name):
    fname = os.path.splitext(xls_name)[0]
    fname = fname.replace('score-', '')
    pars = fname.split('_')
    dpars = {p[:2]: float(p[2:]) for p in pars[1:]}
    dpars['game_type'] = pars[0]
    return dpars

In [5]:
def best_conf(df, player_mean, game_type):
    gdf = df[df['game_type'] == game_type]
    idx = gdf[player_mean].idxmax()
    return gdf.loc[idx]

Here we calculate the average winning rate of a set of players (Exp3, $\varepsilon$-Nash and $\epsilon$-Greedy, in this case)

In [6]:
all_data = []

for f in os.listdir(pth):
    if f.endswith('.xls'):
        dpars = sim_conf(f)
        dmeans = mean_players(os.path.join(pth, f), ['Exp3', 'e-Greedy', 'e-Nash'])
        all_data.append(dict(dpars.items() | dmeans.items()))
        
df_data = pd.DataFrame(all_data)
df_data

Exp3_mean  e-Greedy_mean  e-Nash_mean game_type   ge   ne        pa  \
0   61.198788      62.361818    41.607273   nonstat  0.2  0.2  0.318103   
1   63.481818      62.190303    41.605152   nonstat  0.2  0.2  0.600000   
2   64.518788      62.120000    41.557273   nonstat  0.2  0.2  1.000000   
3   60.393333      62.406970    41.740000   nonstat  0.2  0.2  0.318103   
4   61.529091      62.210606    41.606667   nonstat  0.2  0.2  0.600000   
5   59.441818      62.881818    41.637576   nonstat  0.2  0.2  1.000000   
6   54.363939      62.945455    42.423030   nonstat  0.2  0.2  0.318103   
7   54.519697      63.152424    42.523939   nonstat  0.2  0.2  0.600000   
8   61.535455      58.567879    42.008182   nonstat  0.4  0.2  0.318103   
9   63.615152      58.373939    42.013636   nonstat  0.4  0.2  0.600000   
10  64.916970      58.288182    41.676061   nonstat  0.4  0.2  1.000000   
11  60.682727      58.605152    42.288485   nonstat  0.4  0.2  0.318103   
12  61.796970      58.288182    41.999394   nonstat  0.4  0.2  0.600000   
13  60.040303      58.570909    41.815152   nonstat  0.4  0.2  1.000000   
14  54.572121      58.897576    42.797879   nonstat  0.4  0.2  0.318103   
15  55.035455      59.105455    42.171515   nonstat  0.4  0.2  0.600000   
16  61.147273      62.785758    39.043333   nonstat  0.2  0.4  0.318103   
17  63.378788      62.917273    38.625758   nonstat  0.2  0.4  0.600000   
18  64.594848      62.832121    38.830303   nonstat  0.2  0.4  1.000000   
19  60.495758      63.028485    39.137576   nonstat  0.2  0.4  0.318103   
20  61.626667      62.506667    39.803333   nonstat  0.2  0.4  0.600000   
21  59.568788      62.934545    38.844242   nonstat  0.2  0.4  1.000000   
22  54.425758      63.280606    39.408788   nonstat  0.2  0.4  0.318103   
23  54.809394      63.934848    39.394242   nonstat  0.2  0.4  0.600000   
24  61.638485      58.888485    39.403636   nonstat  0.4  0.4  0.318103   
25  63.843939      58.633333    39.077576   nonstat  0.4  0.4  0.600000   
26  65.000909      58.822121    39.412727   nonstat  0.4  0.4  1.000000   
27  60.861515      58.843030    39.194848   nonstat  0.4  0.4  0.318103   
28  62.120909      58.994848    39.283030   nonstat  0.4  0.4  0.600000   
29  60.172727      58.986970    38.722727   nonstat  0.4  0.4  1.000000   
..        ...            ...          ...       ...  ...  ...       ...   
34  55.531515      52.793030    52.549697      stat  0.2  0.2  1.000000   
35  54.286667      52.655455    52.382424      stat  0.2  0.2  0.318103   
36  54.725758      52.726970    52.646667      stat  0.2  0.2  0.600000   
37  54.045758      52.840000    52.379394      stat  0.2  0.2  1.000000   
38  52.438788      52.696061    52.398788      stat  0.2  0.2  0.318103   
39  52.790000      53.033333    52.406667      stat  0.2  0.2  0.600000   
40  54.555455      52.533030    52.395455      stat  0.4  0.2  0.318103   
41  55.125758      52.670303    52.358485      stat  0.4  0.2  0.600000   
42  55.418485      52.354242    52.537879      stat  0.4  0.2  1.000000   
43  54.424848      52.413939    52.362424      stat  0.4  0.2  0.318103   
44  54.624545      52.515152    52.348485      stat  0.4  0.2  0.600000   
45  54.132424      52.581212    52.475758      stat  0.4  0.2  1.000000   
46  52.674848      52.461212    52.416667      stat  0.4  0.2  0.318103   
47  52.786061      52.582424    52.403030      stat  0.4  0.2  0.600000   
48  54.637576      52.770000    52.810606      stat  0.2  0.4  0.318103   
49  55.514545      53.008485    52.989394      stat  0.2  0.4  0.600000   
50  55.514242      52.803333    52.932727      stat  0.2  0.4  1.000000   
51  54.494242      52.710303    53.038485      stat  0.2  0.4  0.318103   
52  54.764545      52.813939    52.857576      stat  0.2  0.4  0.600000   
53  54.329697      52.623333    53.073636      stat  0.2  0.4  1.000000   
54  52.589394      52.798788    53.087879      stat  0.2  0.4  0.318103   
55  52.854848      52.679091 

We find the best configuration parameters in each game type (stationary or nonstationary) for this set of players by finding the highest win rate of each player in the above table. This will give us the configuration parameters witch leaded each player to its best performance.

These parameters don’t necessarily give the best performance for all players in all cases (this depends of the opponent's configuration), but it's still a reasonable estimate of good parameters for each player.

In [7]:
ns1 = best_conf(df_data, 'Exp3_mean', 'nonstat')[['Exp3_mean', 'pa', 'pg']]
ns2 = best_conf(df_data, 'e-Greedy_mean', 'nonstat')[['e-Greedy_mean', 'ge']]
ns3 = best_conf(df_data, 'e-Nash_mean', 'nonstat')[['e-Nash_mean', 'ne']]
nspd = pd.DataFrame(pd.concat([ns1,ns2,ns3], axis=0)).transpose()

s1 = best_conf(df_data, 'Exp3_mean', 'stat')[['Exp3_mean', 'pa', 'pg']]
s2 = best_conf(df_data, 'e-Greedy_mean', 'stat')[['e-Greedy_mean', 'ge']]
s3 = best_conf(df_data, 'e-Nash_mean', 'stat')[['e-Nash_mean', 'ne']]
spd = pd.DataFrame(pd.concat([s1,s2,s3], axis=0)).transpose()

dr = pd.concat([spd, nspd])
dr.index = ['stat', 'nonstat']
dr

Exp3_mean pa   pg e-Greedy_mean   ge e-Nash_mean   ne
stat      55.5315  1  0.1       53.0333  0.2     53.0879  0.4
nonstat   65.0009  1  0.1       63.9348  0.2     42.7979  0.2